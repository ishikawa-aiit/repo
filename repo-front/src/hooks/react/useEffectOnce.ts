import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
