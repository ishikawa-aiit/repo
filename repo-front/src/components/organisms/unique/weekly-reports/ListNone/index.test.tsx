import { render, screen } from '@testing-library/react'
import React from 'react'

import { ListNone } from '.'

jest.mock('next/link', () => {
  return ({ children, ...rest }: { children: React.ReactElement }) => {
    return React.cloneElement(children, { ...rest })
  }
})

describe('props: isError', () => {
  it('trueの場合、「表示できる週報はありません」と表示される', () => {
    const { container } = render(<ListNone isError={true} />)

    expect(screen.getByText('表示できる週報はありません')).toBeInTheDocument()
    expect(screen.queryByText('作成された週報はありません')).not.toBeInTheDocument()
    expect(screen.queryByTestId('Button')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  describe('falseの場合', () => {
    it('「作成された週報はありません」と表示され、週報作成ページへのリンクボタンが表示される', () => {
      const { container } = render(<ListNone isError={false} />)

      expect(screen.getByText('作成された週報はありません')).toBeInTheDocument()
      expect(screen.queryByText('表示できる週報はありません')).not.toBeInTheDocument()
      expect(screen.getByTestId('Button')).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    it('指定のパスとリンクのhref属性が一致する', () => {
      const { container } = render(<ListNone isError={false} />)

      expect(screen.getByTestId('Button')).toHaveAttribute('href', '/weekly-reports/create')
      expect(container).toMatchSnapshot()
    })
  })
})
