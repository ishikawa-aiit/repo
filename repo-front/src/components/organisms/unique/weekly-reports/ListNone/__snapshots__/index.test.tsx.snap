// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props: isError falseの場合 「作成された週報はありません」と表示され、週報作成ページへのリンクボタンが表示される 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 6px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-1 {
  font-weight: bold;
  font-size: xl;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

.emotion-4 {
  width: 5px;
}

<div>
  <div
    class="chakra-stack emotion-0"
  >
    <p
      class="chakra-text emotion-1"
      data-cy="ListNone"
      data-testid="ListNone"
    >
      作成された週報はありません
    </p>
    <a
      class="chakra-button emotion-2"
      data-cy="Button"
      data-testid="Button"
      href="/weekly-reports/create"
    >
      <span
        class="chakra-button__icon emotion-3"
      >
        <span
          aria-hidden="true"
          class="emotion-4"
          focusable="false"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
      </span>
      週報を作成する
    </a>
  </div>
</div>
`;

exports[`props: isError falseの場合 指定のパスとリンクのhref属性が一致する 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 6px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-1 {
  font-weight: bold;
  font-size: xl;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-margin-end: 0.5rem;
  margin-inline-end: 0.5rem;
}

.emotion-4 {
  width: 5px;
}

<div>
  <div
    class="chakra-stack emotion-0"
  >
    <p
      class="chakra-text emotion-1"
      data-cy="ListNone"
      data-testid="ListNone"
    >
      作成された週報はありません
    </p>
    <a
      class="chakra-button emotion-2"
      data-cy="Button"
      data-testid="Button"
      href="/weekly-reports/create"
    >
      <span
        class="chakra-button__icon emotion-3"
      >
        <span
          aria-hidden="true"
          class="emotion-4"
          focusable="false"
        >
          <svg
            aria-hidden="true"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
      </span>
      週報を作成する
    </a>
  </div>
</div>
`;

exports[`props: isError trueの場合、「表示できる週報はありません」と表示される 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: 6px;
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-1 {
  font-weight: bold;
  font-size: xl;
}

<div>
  <div
    class="chakra-stack emotion-0"
  >
    <p
      class="chakra-text emotion-1"
      data-cy="ListNoneIsError"
      data-testid="ListNoneIsError"
    >
      表示できる週報はありません
    </p>
  </div>
</div>
`;
